File: server/assets/solutions/ohtu-avoin-2022/part5/python/tennis/src/tennis_game.py
class TennisGame:
    def __init__(self, player1_name, player2_name):
        self.player1_name = player1_name
        self.player2_name = player2_name
        self.score_1 = 0
        self.score_2 = 0

        self.numbers = {
          0: "Love",
          1: "Fifteen",
          2: "Thirty",
          3: "Forty"
        }

    def won_point(self, player_name):
        if player_name == "player1":
            self.score_1 = self.score_1 + 1
        else:
            self.score_2 = self.score_2 + 1

    def tie_score(self):
        if self.score_1 < 4:
            return  f"{self.numbers[self.score_1]}-All"
        else:
           return "Deuce"

    def possible_winning_score(self):
        difference = self.score_1 - self.score_2
        player = "player1" if difference > 0 else "player2"

        if abs(difference) == 1:
            return f"Advantage {player}"

        return f"Win for {player}"
      
    def not_decisive_score(self):
      return f"{self.numbers[self.score_1]}-{self.numbers[self.score_2]}"

    def get_score(self):
        if self.score_1 == self.score_2:
            return self.tie_score()
        elif self.score_1 >= 4 or self.score_2 >= 4:
            return self.possible_winning_score()
        else:
            return self.not_decisive_score()

File: server/assets/solutions/ohtu-avoin-2021/part4/java/Tennis/src/main/java/ohtu/TennisGame.java
package ohtu;

import java.util.HashMap;
import java.util.Map;

public class TennisGame {

    private int score1 = 0;
    private int score2 = 0;
    private final String player1;
    private final String player2;
    private Map<Integer, String> numberToScore = new HashMap<>();
    
    public TennisGame(String player1, String player2) {
        this.player1 = player1;
        this.player2 = player2;
        numberToScore.put(0, "Love");
        numberToScore.put(1, "Fifteen");
        numberToScore.put(2, "Thirty");
        numberToScore.put(3, "Forty");
    }

    public void wonPoint(String player) {
        if (player.equals(player1)) {
            score1 += 1;
        } else {
            score2 += 1;
        }
    }

    public String getScore() {
        if (isEven()) {
            return evenScore();
        } else if (isPossibleWin()) {
            return possibleWinScore();
        }

        return leadScore();
    }

    private boolean isEven() {
        return score1 == score2;
    }

    private boolean isPossibleWin() {
        return score1 >= 4 || score2 >= 4;
    }
    
    private String evenScore() {  
        if (score1 > 3) {
            return "Deuce";
        }
        return numberToScore.get(score1)+"-All";
    }
    
    private String possibleWinScore() {
        String better = score1 > score2 ? player1 : player2;
        String result =  Math.abs(score1 - score2) > 1 ?  "Win for" : "Advantage";
        
        return result + " "+ better;
    }
    
    private String leadScore() {
        return numberToScore.get(score1)+"-"+numberToScore.get(score2);
    }    
}
